### PR TITLE
Subclassed reshape

### LINKED ISSUE #15563: Preserve subclass family on reshape operations
Subclassed Series and DataFrames with custom `_constructor`, `_constructor_sliced`, and `_constructor_expanddim` methods return pandas Series and DataFrame objects on reshape operations:

```python
>>> import numpy as np, pandas as pd
>>> import pandas.util.testing as tm
>>> df = tm.SubclassedDataFrame(np.random.random((4,5)))
>>> df
          0         1         2         3         4
0  0.151944  0.948561  0.639122  0.718071  0.296193
1  0.004606  0.978378  0.827614  0.479320  0.495945
2  0.158874  0.535476  0.173194  0.065292  0.365851
3  0.772392  0.105038  0.671064  0.164165  0.795803
>>> type(df)
<class 'pandas.util.testing.SubclassedDataFrame'>
```

Slicing operations use `_constructor_sliced`:
```python
>>> type(df[0])
<class 'pandas.util.testing.SubclassedSeries'>
```

Reshape operations return regular pandas objects:
```python
>>> type(df.stack())
<class 'pandas.core.series.Series'>
```

It would be great if this returned a `SublcassedSeries`. Same goes for unstack and pivot.

### PR DESCRIPTION
**Pull request by delgadom**
- [x] closes #15563
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

Reshape operations in pandas/core/reshape.py modified to use `obj._constructor`, `obj._constructor_sliced`, and `obj._constructor_expanddim` for Series and DataFrame. This preserves a subclass family on `stack`, `unstack`, and `pivot` operations.

### CODE REVIEW

**Comment by jreback** [comment]
can you also have a look over: http://pandas-docs.github.io/pandas-docs-travis/internals.html#subclassing-pandas-data-structures

and see if anything needs amendment

File: doc/source/whatsnew/v0.9.1.txt [hunk 1]
```diff
@@ -105,6 +105,8 @@ New features
   - DataFrame.drop now supports non-unique indexes (:issue:`2101`)
   - Panel.shift now supports negative periods (:issue:`2164`)
   - DataFrame now support unary ~ operator (:issue:`2110`)
+  - `stack`, `unstack`, and `pivot` operations now preserve subclass family
```
Comment (jreback) [inline doc/source/whatsnew/v0.9.1.txt]: move to 0.20.0

File: pandas/core/reshape.py [hunk 2]
```diff
@@ -410,13 +413,24 @@ def unstack(obj, level, fill_value=None):
 
     if isinstance(obj, DataFrame):
         if isinstance(obj.index, MultiIndex):
-            return _unstack_frame(obj, level, fill_value=fill_value)
+            unstacked = _unstack_frame(obj, level, fill_value=fill_value)
+        else:
```
Comment (jreback) [inline pandas/core/reshape.py]: hmm I think this should be done *inside* the ``.stack`` and ``_unstack_frame`` functions no?

File: pandas/core/reshape.py [hunk 3]
```diff
@@ -410,13 +413,24 @@ def unstack(obj, level, fill_value=None):
...
         else:
-            return obj.T.stack(dropna=False)
+            return obj._constructor(unstacked)
+
     else:
         unstacker = _Unstacker(obj.values, obj.index, level=level,
                                fill_value=fill_value)
-        return unstacker.get_result()
+        unstacked = unstacker.get_result()
+
```
Comment (jreback) [inline pandas/core/reshape.py]: same here, this now has logic in 2 places, the creation itself of the unstacked result, then the return class. should be a part of ``get_result`` (it certainly could be another method that is *called* by ``get_result``) though

File: pandas/core/reshape.py [hunk 4]
```diff
@@ -515,7 +529,7 @@ def factorize(index):
         mask = notnull(new_values)
         new_values = new_values[mask]
         new_index = new_index[mask]
-    return Series(new_values, index=new_index)
+    return frame._constructor_sliced(new_values, index=new_index)
```
Comment (jreback) [inline pandas/core/reshape.py]: so this is good

File: pandas/tests/frame/test_subclass.py [hunk 5]
```diff
@@ -125,6 +125,144 @@ def test_indexing_sliced(self):
         tm.assert_series_equal(res, exp)
         tm.assertIsInstance(res, tm.SubclassedSeries)
 
+    def test_subclass_stack(self):
+        df = tm.SubclassedDataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
```
Comment (jreback) [inline pandas/tests/frame/test_subclass.py]: can you add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
thanks @delgadom 

reshape could certainly use some tlc!

**Comment by delgadom** [comment]
Thanks @jreback - I'm working on your suggestions.

Keeping this clean is a bit tough, especially because unstacking is sometimes handled by `_unstack_multiple` and sometimes by `_Unstacker`. I'd modify `_unstack_multiple` and `_Unstacker`, except `_Unstacker` accepts a numpy array and returns a `DataArray`, so there's no way to access the `_constructor*` properties from within this class.

To me this class seems really nasty and should be handled by functions rather than a class (would you ever use `_Unstacker` except to get an unstacked object?) but originally was tentative in making changes. Do you think this should be re-written? At the very least it seems as though `_Unstacker` should be able to handle `DataFrame` arguments.

Happy to take a stab if you agree.

**Comment by jreback** [comment]
@delgadom the reshaping code is certainly in need of refactoring to make things easier :>

certainly take a stab at it. Yes the ``_Unstacker`` class is merely a helper class, so you can change things if you need (not public).

You *might* want to do a refactor *before* this PR to make it easier to test / debug. Though if its relatively simple go ahead here.

**Comment by delgadom** [comment]
Yeah I think I'll refactor and then create a new PR. Thanks for the help!

**Comment by jreback** [comment]
ok!

**Pull request closed without merging**